File: src/python/Rectangle2.py
from Vector2 import Vector2

class Rectangle2(object):
    def __init__(self, min_point, max_point):
        self._min_x = min_point.x
        self._min_y = min_point.y
        self._max_x = max_point.x
        self._max_y = max_point.y

    def get_min_x(self):
        return self._min_x

    min_x = property(get_min_x)

    def get_min_y(self):
        return self._min_y

    min_y = property(get_min_y)

    def get_max_x(self):
        return self._max_x

    max_x = property(get_max_x)

    def get_max_y(self):
        return self._max_y

    max_y = property(get_max_y)

    def get_min_point(self):
        return Vector2(self._min_x, self._min_y)

    min_point = property(get_min_point)

    def get_max_point(self):
        return Vector2(self._max_x, self._max_y)

    max_point = property(get_max_point)

    def get_bounding_box(self):
        return self

    bounding_box = property(get_bounding_box)

    def contains_point(self, point):
        return (point.x >= self.min_x and point.x <= self.max_x
                and point.y >= self.min_y and point.y <= self.max_y)

File: src/cpp/siblings/rectangle_2.cpp
#include "rectangle_2.hpp"
#include <algorithm>
#include <ostream>

namespace siblings {
    rectangle_2::rectangle_2() { }

    rectangle_2::rectangle_2(const vector_2& min, const vector_2& max)
        : min_(std::min(min.x(), max.x()), std::min(min.y(), max.y())),
          max_(std::max(min.x(), max.x()), std::max(min.y(), max.y()))
    { }

    rectangle_2::rectangle_2(real min_x, real min_y, real max_x, real max_y)
        : min_(std::min(min_x, max_x), std::min(min_y, max_y)),
          max_(std::max(min_x, max_x), std::max(min_y, max_y))
    { }

    const vector_2& rectangle_2::min() const { return min_; }
    const vector_2& rectangle_2::max() const { return max_; }

    std::ostream& operator<<(std::ostream& out, const rectangle_2& r)
    {
        return out << "[" << r.min().x() << ", " << r.min().y() << "; "
                   << r.max().x() << ", " << r.max().y() << "]";
    }
}
